internal/impl/sftp/input.go

```diff
@@ -173,177 +176,243 @@ func newSFTPReaderFromParsed(conf *service.ParsedConfig, mgr *service.Resources)
-				_ = s.pathProvider.Ack(ctx, nextPath, nil)
-			} else {
-				// Otherwise we "nack" it with the error as we'll want to
-				// reprocess it again later.
-				_ = s.pathProvider.Ack(ctx, nextPath, err)
-			}
-		} else {
-			break
-		}
+	if s.client != nil {
+		return s.client, func() {}, nil
 	}
 
-	details := service.NewScannerSourceDetails()
-	details.SetName(nextPath)
-	if s.scanner, err = s.scannerCtor.Create(file, func(ctx context.Context, aErr error) (outErr error) {
-		_ = s.pathProvider.Ack(ctx, nextPath, aErr)
-		if aErr != nil {
-			return nil
-		}
-		if s.deleteOnFinish {
-			s.scannerMut.Lock()
-			client := s.client
-			if client == nil {
-				if client, outErr = s.creds.GetClient(s.mgr.FS(), s.address); outErr != nil {
-					outErr = fmt.Errorf("obtain private client: %w", outErr)
-				}
-				defer func() {
-					_ = client.Close()
-				}()
-			}
-			if outErr == nil {
-				if outErr = client.Remove(nextPath); outErr != nil {
-					outErr = fmt.Errorf("remove %v: %w", nextPath, outErr)
-				}
-			}
-			s.scannerMut.Unlock()
+	client, err := s.creds.GetClient(s.mgr.FS(), s.address)
+	if err != nil {
+		return nil, nil, fmt.Errorf("initializing SFTP client: %w", err)
+	}
+
+	if s.clientClosedForever {
+		release := func() { s.closeClient(false) }
+		return client, release, nil
+	}
+
+	if s.clientClosedForever {
+		client, err := s.creds.GetClient(s.mgr.FS(), s.address)
+		if err != nil {
+			return nil, nil, fmt.Errorf("initializing SFTP client: %w", err)
 		}
-		return
-	}, details); err != nil {
-		_ = file.Close()
-		_ = s.pathProvider.Ack(ctx, nextPath, err)
-		return err
+		return client, func() { s.closeClient(true) }, nil
 	}
-	s.currentPath = nextPath
 
-	s.log.Debugf("Consuming from file '%v'", nextPath)
-	return
+	if s.client == nil {
+		var err error
+		if s.client, err = s.creds.GetClient(s.mgr.FS(), s.address); err != nil {
+			return nil, nil, fmt.Errorf("initializing SFTP client: %w", err)
+		}
+	}
+	return s.client, func() {}, nil
 }
 
 func (s *sftpReader) ReadBatch(ctx context.Context) (service.MessageBatch, service.AckFunc, error) {
-	s.scannerMut.Lock()
-	scanner := s.scanner
-	client := s.client
-	currentPath := s.currentPath
-	s.scannerMut.Unlock()
+	parts, codecAckFn, err := s.tryReadBatch(ctx)
+	if err != nil {
+		if errors.Is(err, sftp.ErrSSHFxConnectionLost) {
+			s.stateLock.Lock()
+			s.closeScanner(ctx)
+			s.closeClient(false)
+			s.stateLock.Unlock()
+			err = service.ErrNotConnected
+		}
+		return nil, nil, err
+	}
+	return parts, codecAckFn, nil
+}
 
-	if scanner == nil || client == nil {
-		return nil, nil, service.ErrNotConnected
+func (s *sftpReader) tryReadBatch(ctx context.Context) (service.MessageBatch, service.AckFunc, error) {
+	if err := s.initScanner(ctx); err != nil {
+		return nil, nil, err
 	}
 
-	parts, codecAckFn, err := scanner.NextBatch(ctx)
+	s.stateLock.Lock()
+	defer s.stateLock.Unlock()
+
+	parts, codecAckFn, err := s.scanner.NextBatch(ctx)
```

Comment:
I'm not sure it's actually possible to be hit in practice but there's a race condition here, as the lock around `s.scanner` is yielded within `initScanner` before the lock is re-acquired, so there's opportunity for another goroutine to set `s.scanner` to `nil` in that gap. If you want to avoid this then have `initScanner` return the scanner pointer and use that reference.

Reply:
Wow good catch. I'll do this